Performance.timing should be a PerformanceTiming object

aniebietafia opened the pull request:
According to the [specs](https://w3c.github.io/navigation-timing/#extensions-to-the-performance-interface) `timing` member should be a `PerformanceTiming object`. 
- Added the `PerformanceTiming` interface in `components/script_bindings/webidls/`
- Updated the `Performance.webidl` from `PerformanceNavigationTiming` to `PerformanceTiming`.
```bash
[Exposed=Window]
partial interface Performance {
  [SameObject]
  readonly attribute PerformanceTiming timing;
  [SameObject]
  readonly attribute PerformanceNavigation navigation;
```
- Implemented the attributes in the `PerformanceTiming` object in `performancetiming.rs`.
- Added the `timing` and `navigation` to the `Performance` struct in `performance.rs`.
- Updated `mod.rs` with `performancetiming`.
- called the update functions from `performancetiming.rs` in `document.rs`.

Testing: *Describe how this pull request is tested or why it doesn't require tests*
Testing happens in `tests/wpt/tests/navigation-timing` to see what happens.

Fixes:
This PR fixes #36171

aniebietafia committed:
Replace some uses of window.global() with upcast.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

aniebietafia committed:
Performance.timing should be a PerformanceTiming object.

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

aniebietafia commented:
I have seen the failed cases and will set out to fix them. Had to push these changes to see what's failing as Iam using an 8GB RAM and it is damn too slow to run the build and test cases.

TG199 commented:
> I have seen the failed cases and will set out to fix them. Had to push these changes to see what's failing as Iam using an 8GB RAM and it is damn too slow to run the build and test cases.

About the RAM, have you considered shallow clone? Had similar issues when starting and went with shallow clone. The Servo project is really big. 
How to go about it:
"Servo’s main repo is pretty big! If you have an unreliable network connection or limited disk space, consider making a shallow clone."
 https://book.servo.org/getting-servo.html#:~:text=Servo%E2%80%99s%20main%20repo%20is%20pretty%20big!%20If%20you%20have%20an%20unreliable%20network%20connection%20or%20limited%20disk%20space%2C%20consider%20making%20a%20shallow%20clone.

https://github.blog/open-source/git/get-up-to-speed-with-partial-clone-and-shallow-clone/#:~:text=git%20clone%20%2D%2Ddepth%3D1%20%3Curl%3E

jdm commented:
@aniebietafia Are you still working on this?

aniebietafia commented:
@jdm yes I am

aniebietafia commented on components/script/dom/document.rs:
```diff
@@ -1093,8 +1108,13 @@ impl Document {
                     ));
                 }
                 update_with_current_instant(&self.dom_complete);
+                timing.update_dom_complete(now);
+                timing.update_load_event_start(now);
+
+                self.upcast::<EventTarget>()
+                    .fire_event(atom!("load"), can_gc);
+                timing.update_load_event_end((*performance.Now()).floor() as u64);
```
Ouch! I was just testing something out and should have gotten rid of the `fire_event`. Should I add check ensure events are only fired if they haven't been fired before or getting rid of it should be just fine?

jdm replied:
We should move the code that times the events to where the events are fired:
* https://github.com/servo/servo/blob/main/components/script/dom/document.rs#L3038
* https://github.com/servo/servo/blob/main/components/script/dom/document.rs#L2774

aniebietafia replied:
I have refactored to move the code to where the events are fired.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

jdm commented on components/script/dom/document.rs:
```diff
@@ -2778,6 +2793,10 @@ impl Document {
                 );
                 event.set_trusted(true);
 
+                timing.update_load_event_start(start_time);
+                window.upcast::<EventTarget>().dispatch_event(&event, CanGc::note());
+                timing.update_load_event_end((*performance.Now()).floor() as u64);
```
We already dispatch the event below in `window.dispatch_event_with_target_override(&event, CanGc::note());`, so we should update the timing information based on that event.

jdm commented on components/script/dom/document.rs:
```diff
@@ -2769,6 +2779,11 @@ impl Document {
+
+                let performance = window.Performance();
+                let timing = performance.Timing();
+                let start_time = (*performance.Now()).floor() as u64;
```
This timestamp should be computed right before we dispatch the event.

jdm commented on components/script/dom/document.rs:
```diff
@@ -3025,20 +3044,34 @@ impl Document {
+            let timing = performance.Timing();
+            let start_time = (*performance.Now()).floor() as u64;
+
+            let event = Event::new(
+                window.upcast(),
+                atom!("DOMContentLoaded"),
+                EventBubbles::Bubbles,
+                EventCancelable::NotCancelable,
+                CanGc::note(),
             );
+            event.set_trusted(true);
+
+            timing.update_dom_content_loaded_event_start(start_time);
+            window.upcast::<EventTarget>().dispatch_event(&event, CanGc::note());
```
We should use the previous code that was here instead:
```rust
document.upcast::<EventTarget>().fire_bubbling_event(atom!("DOMContentLoaded"), CanGc::note());
```

jdm commented on components/script/dom/performance.rs:
```diff
@@ -148,10 +148,19 @@ pub(crate) struct Performance {
+    fn new_inherited(global: &GlobalScope, time_origin: CrossProcessInstant) -> Performance {
+        let nav_start = Self::time_origin_to_millis(time_origin);
+        let timing = PerformanceTiming::new(global, nav_start);
+        let navigation = PerformanceNavigation::new(global, CanGc::note());
```
Please add a `can_gc: CanGc` argument to this method so callers pass it in.
```suggestion
        let navigation = PerformanceNavigation::new(global, can_gc);
```

jdm commented on components/script/dom/performancetiming.rs:
```diff
@@ -0,0 +1,282 @@
+
+#[allow(dead_code)]
+impl PerformanceTiming {
+    pub fn new_inherited(config: PerformanceTimingBuilder) -> PerformanceTiming {
```
```suggestion
    fn new_inherited(config: PerformanceTimingBuilder) -> PerformanceTiming {
```

jdm commented on components/script/dom/performancetiming.rs:
```diff
@@ -0,0 +1,282 @@
+use crate::script_runtime::CanGc;
+
+#[dom_struct]
+pub struct PerformanceTiming {
```
Please use `pub(crate)` instead of `pub` in this PR, since we want to limit types that are exposed outside of this crate by default.
```suggestion
pub(crate) struct PerformanceTiming {
```

jdm commented on components/script/dom/performancetiming.rs:
```diff
@@ -0,0 +1,282 @@
+    }
+
+    pub fn new(global: &GlobalScope, navigation_start: u64) -> DomRoot<PerformanceTiming> {
+        let config = PerformanceTimingBuilder {
+            navigation_start,
+            ..Default::default()
+        };
```
This builder type feels unnecessary right now, since we only pass in one value and there are no other users of it. Can we remove it and just set the field directly?

jdm commented on components/script/dom/document.rs:
```diff
@@ -3025,20 +3044,34 @@ impl Document {
             "Complete before DOMContentLoaded?"
         );
 
-        update_with_current_instant(&self.dom_content_loaded_event_start);
```
We lost this timestamp. Can you move it right before the event dispatch?

jdm reviewed (changes requested):
Great work! Just some small changes requested to make sure we're not changing behaviour.

aniebietafia committed:
Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object.

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

Performance.timing should be a PerformanceTiming object

Signed-off-by: aniebietafia <aniebietafia87@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

aniebietafia commented:
All changes have been successfully implemented and is ready for review.

jdm commented on components/script/dom/performancetiming.rs:
```diff
@@ -0,0 +1,253 @@
+        )
+    }
+
+    pub fn update_unload_event_start(&self, value: u64) {
```
pub(crate) for all of these methods.

jdm commented:
You should check the results of `./mach test-wpt tests/wpt/tests/navigation-timing` after these changes, too.

aniebietafia committed:
Merge branch 'main' into outreachy-intern

Signed-off-by: Aniebiet Afia <54312052+aniebietafia@users.noreply.github.com>

aniebietafia committed:
Merge branch 'main' into outreachy-intern

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

aniebietafia commented:
@jdm I ran into some challenges recently and as a result, I might not be focused to complete this. It is totally ok if someone else can complete it. Thank you.

jdm commented:
Closing in favour of #46990.
